src/replica/replica_throttle.cpp

```diff
@@ -14,40 +14,49 @@
+        int64_t delay_ms = 0;                                                                      \
+        auto type = _##op_type##_##throttling_type##_throttling_controller.control(                \
+            request->header->client.timeout_ms, request_units, delay_ms);                          \
+        if (type != throttling_controller::PASS) {                                                 \
+            if (type == throttling_controller::DELAY) {                                            \
+                tasking::enqueue(                                                                  \
+                    LPC_##op_type##_THROTTLING_DELAY,                                              \
+                    &_tracker,                                                                     \
+                    [ this, req = message_ptr(request) ]() { on_client_##op_type(req, true); },    \
+                    get_gpid().thread_hash(),                                                      \
+                    std::chrono::milliseconds(delay_ms));                                          \
+                _counter_recent_##op_type##_throttling_delay_count->increment();                   \
+            } else { /** type == throttling_controller::REJECT **/                                 \
+                if (delay_ms > 0) {                                                                \
+                    tasking::enqueue(LPC_##op_type##_THROTTLING_DELAY,                             \
+                                     &_tracker,                                                    \
+                                     [ this, req = message_ptr(request) ]() {                      \
+                                         response_client_##op_type(req, ERR_BUSY);                 \
+                                     },                                                            \
+                                     get_gpid().thread_hash(),                                     \
+                                     std::chrono::milliseconds(delay_ms));                         \
+                } else {                                                                           \
+                    response_client_##op_type(request, ERR_BUSY);                                  \
+                }                                                                                  \
+                _counter_recent_##op_type##_throttling_reject_count->increment();                  \
+            }                                                                                      \
+            return true;                                                                           \
+        }                                                                                          \
+    } while (0)
+
+bool replica::throttle_write_request(message_ex *request)
 {
-    if (!controller.enabled()) {
```

Comment:
I recommend remaining check controller enabled in function `throttle_write_request` and `throttle_read_request` to avoid calling function `control`, especially for read reuqest, most of read requests don't have throttle control.

Reply:
I didn't remove it.  I just move this check into `src/utils/throttling_controller.cpp`.

Reply:
I knew that check is in function `control`, my suggestion is not moving the check in function `control`, remaining it in function `throttle_write_request` and `throttle_read_request` to avoid calling another function, especially for read requests.

Reply:
There are two contrllers for read and write operations, which are size throttling and qps throttling. we should call enable() function for each of them.
And I think It's better to move enable() into throttling_controller to simplify the usage of it.

Reply:
I got your point. How about moving this check in macro THROTTLE_REQUEST? It can avoid calling another function and simplify for usage.

Reply:
I don't think it's a good idea. We shouldn't make the `enable` of `throttling_controller` visiable. In other words,  for the user of `throttling_controller`,  it is not necessary to know whether `enable` is set or not. He only needs to call the function `control` and get the result which means whether it is passed or not.

Reply:
I don't think it's a good idea. We shouldn't make the `enable` of `throttling_controller` visiable. In other words,  for the user of `throttling_controller`,  it is not necessary to know whether `enable` is set or not. He only needs to call the function `control` and get the result which means whether it is passed or not.